File: src/layer1_wasm/numpy-28287/repro.ts
// Vivarium Layer 1 reproduction — numpy/numpy#28287.
//
// `timedelta64` ordering is non-transitive when one of the values uses
// the generic unit. With:
//   x = np.timedelta64(1, "ms")
//   y = np.timedelta64(2)         # generic unit
//   z = np.timedelta64(5, "ns")
// NumPy reports x < y and y < z but x > z — a transitivity violation.
//
// Verdict semantics (per ADR-0008 / contract v1):
//   - "pass" — the bug REPRODUCES (numpy reports the non-transitive
//     ordering on the build Pyodide ships).
//   - "fail" — the bug does NOT reproduce (or the runtime errored).

import { loadVivariumPyodide } from "../_shared/loader.js";
import {
  setResult,
  setVerdict,
  type VivariumResultV1,
} from "../_shared/verdict.js";

const REPRO_CODE = `
import sys
import numpy as np

x = np.timedelta64(1, "ms")
y = np.timedelta64(2)
z = np.timedelta64(5, "ns")

x_lt_y = bool(x < y)
y_lt_z = bool(y < z)
x_lt_z = bool(x < z)

{
    "numpy_version": np.__version__,
    "python_version": sys.version.split()[0],
    "x_lt_y": x_lt_y,
    "y_lt_z": y_lt_z,
    "x_lt_z": x_lt_z,
    "transitivity_violated": x_lt_y and y_lt_z and not x_lt_z,
}
`.trim();

interface ReproOutput {
  numpy_version: string;
  python_version: string;
  x_lt_y: boolean;
  y_lt_z: boolean;
  x_lt_z: boolean;
  transitivity_violated: boolean;
}

interface PyodideRuntime {
  runPythonAsync(code: string): Promise<{
    toJs(opts: { dict_converter: typeof Object.fromEntries }): ReproOutput;
    destroy?(): void;
  }>;
}

const outputEl = document.getElementById("output");
const metaEl = document.getElementById("meta");
const reproCodeEl = document.getElementById("repro-code");

if (!outputEl || !metaEl || !reproCodeEl) {
  throw new Error(
    "numpy-28287: missing required DOM elements (#output, #meta, #repro-code).",
  );
}

reproCodeEl.textContent = REPRO_CODE;

const startedAt = new Date();

try {
  // numpy is shipped with the Pyodide distribution but must be installed
  // explicitly via `loadPackage` / the `packages` option — it is not
  // imported at runtime startup.
  const { pyodide, version } = await loadVivariumPyodide({
    packages: ["numpy"],
    pendingText: "Loading Pyodide runtime and numpy…",
  });

  setVerdict("pending", "Running reproduction script…");
  const runtime = pyodide as PyodideRuntime;
  const proxy = await runtime.runPythonAsync(REPRO_CODE);
  const result = proxy.toJs({ dict_converter: Object.fromEntries });
  proxy.destroy?.();

  metaEl.textContent =
    `numpy ${result.numpy_version} on Python ${result.python_version} ` +
    `via Pyodide v${version}.`;
  outputEl.textContent = JSON.stringify(result, null, 2);

  if (result.transitivity_violated) {
    setVerdict(
      "pass",
      "reproduction succeeded — timedelta64 ordering is non-transitive (x < y < z but x ≥ z).",
    );
  } else {
    setVerdict(
      "fail",
      "reproduction failed — timedelta64 ordering is transitive in this numpy build.",
    );
  }

  const finishedAt = new Date();
  const envelope: VivariumResultV1 = {
    contract: "v1",
    bug: {
      project: "numpy",
      issue: 28287,
      upstream_url: "https://github.com/numpy/numpy/issues/28287",
    },
    runtime: {
      name: "pyodide",
      version,
      extras: {
        python: result.python_version,
        numpy: result.numpy_version,
      },
    },
    result: {
      x_lt_y: result.x_lt_y,
      y_lt_z: result.y_lt_z,
      x_lt_z: result.x_lt_z,
      transitivity_violated: result.transitivity_violated,
    },
    timing: {
      started_at: startedAt.toISOString(),
      finished_at: finishedAt.toISOString(),
      duration_ms: finishedAt.getTime() - startedAt.getTime(),
    },
  };
  setResult(envelope);
} catch (err: unknown) {
  console.error(err);
  const errAny = err as { stack?: string; message?: string } | null;
  outputEl.textContent =
    (errAny && (errAny.stack ?? errAny.message)) ?? String(err);
  // `loadVivariumPyodide` already sets "fail" on load-time errors. Cover
  // the case where the runtime loaded but the reproduction itself errored.
  if (globalThis.__VIVARIUM_VERDICT__ !== "fail") {
    setVerdict(
      "fail",
      `reproduction failed — runtime error: ${errAny?.message ?? String(err)}`,
    );
  }
}

File: src/layer1_wasm/numpy-28287/repro.py
# /// script
# requires-python = ">=3.13"
# dependencies = [
#   "numpy==2.2.5",
# ]
# ///
"""Vivarium Layer 1 reproduction — numpy/numpy#28287, native variant.

Mirrors the script that runs in `repro.ts` (under Pyodide) so a
contributor can re-verify the bug against a real CPython interpreter
+ a real NumPy build:

    mise install                                                  # one-time
    mise exec uv -- uv run src/layer1_wasm/numpy-28287/repro.py

PEP 723 inline metadata pins numpy to **2.2.5** — the exact version
Pyodide v0.29.3 ships — so the page and this CLI exercise the same
code path.

Prints `pass` if the bug REPRODUCES (`timedelta64` ordering is
non-transitive across the generic unit), `fail` otherwise. Exit
code: 0 on `pass`, 1 on `fail`.
"""

import json
import sys

import numpy as np

x = np.timedelta64(1, "ms")
y = np.timedelta64(2)
z = np.timedelta64(5, "ns")

x_lt_y = bool(x < y)
y_lt_z = bool(y < z)
x_lt_z = bool(x < z)
transitivity_violated = x_lt_y and y_lt_z and not x_lt_z

result = {
    "numpy_version": np.__version__,
    "python_version": sys.version.split()[0],
    "x_lt_y": x_lt_y,
    "y_lt_z": y_lt_z,
    "x_lt_z": x_lt_z,
    "transitivity_violated": transitivity_violated,
    "reproduced": transitivity_violated,
}

print(json.dumps(result, indent=2))

if transitivity_violated:
    print(
        "verdict=pass — timedelta64 ordering is non-transitive "
        "(x < y < z but x ≥ z)",
        file=sys.stderr,
    )
    sys.exit(0)
else:
    print(
        "verdict=fail — timedelta64 ordering is transitive "
        "(likely fixed upstream)",
        file=sys.stderr,
    )
    sys.exit(1)
